services/spar/src/Spar/API.hs

```diff
@@ -0,0 +1,246 @@
+
+app :: Env -> Application
+app ctx = SAML.setHttpCachePolicy
+        $ serve (Proxy @API) (enter (NT (SAML.nt @Spar ctx)) api :: Server API)
+
+type API = "i" :> "status" :> GetNoContent '[JSON] NoContent
+      :<|> "sso" :> "api-docs" :> Get '[JSON] Swagger
+      :<|> APIMeta
+      :<|> APIAuthReq
+      :<|> APIAuthResp
+      :<|> IdpGet
+      :<|> IdpCreate
+      :<|> IdpDelete
+
+type APIMeta     = "sso" :> "metainfo" :> SAML.APIMeta
+type APIAuthReq  = "sso" :> "initiate-login" :> SAML.APIAuthReq
+type APIAuthResp = "sso" :> "finalize-login" :> SAML.APIAuthResp
+
+type IdpGet     = Header "Z-User" Brig.UserId :> "sso" :> "identity-providers" :> Capture "id" SAML.IdPId :> Get '[JSON] IdP
+type IdpCreate  = Header "Z-User" Brig.UserId :> "sso" :> "identity-providers" :> ReqBody '[JSON] NewIdP :> PostCreated '[JSON] IdP
+type IdpDelete  = Header "Z-User" Brig.UserId :> "sso" :> "identity-providers" :> Capture "id" SAML.IdPId :> DeleteNoContent '[JSON] NoContent
+
+
+api :: ServerT API Spar
+api =  pure NoContent
+  :<|> pure (toSwagger (Proxy @API))
+  :<|> SAML.meta appName (Proxy @API) (Proxy @APIAuthResp)
+  :<|> SAML.authreq
+  :<|> SAML.authresp onSuccess
+  :<|> idpGet
+  :<|> idpCreate
+  :<|> idpDelete
+
+appName :: ST
+appName = "spar"
+
+onSuccess :: HasCallStack => SAML.UserRef -> Spar (SetCookie, URI.URI)
+onSuccess uid = forwardBrigLogin =<< maybe (createUser uid) pure =<< getUser uid
+
+type ZUsr = Maybe Brig.UserId
+
+idpGet :: ZUsr -> SAML.IdPId -> Spar IdP
+idpGet zusr idpid = withDebugLog "idpGet" (Just . show . (^. SAML.idpId)) $ do
+  authorizeIdP zusr =<< SAML.getIdPConfig idpid
+
+idpDelete :: ZUsr -> SAML.IdPId -> Spar NoContent
+idpDelete zusr idpid = withDebugLog "idpDelete" (const Nothing) $ do
+    idp <- SAML.getIdPConfig idpid
+    void $ authorizeIdP zusr idp
+    wrapMonadClient $ Data.deleteIdPConfig idpid (idp ^. SAML.idpIssuer) (idp ^. SAML.idpExtraInfo)
+    return NoContent
+
+-- We generate a new UUID for each IdP used as IdPConfig's path, thereby ensuring uniqueness
+idpCreate :: ( SAML.SP m, SAML.SPStoreIdP m, SAML.ConfigExtra m ~ Brig.TeamId
+             , MonadError ServantErr m
+             , Brig.MonadSparToBrig m
+             )
+          => ZUsr -> NewIdP -> m IdP
+idpCreate zusr newIdP = withDebugLog "idpCreate" (Just . show . (^. SAML.idpId)) $ do
+  teamid <- getZUsrTeam zusr
+  idp <- initializeIdP newIdP teamid
+  SAML.storeIdPConfig idp
+  pure idp
+
+withDebugLog :: SAML.SP m => String -> (a -> Maybe String) -> m a -> m a
+withDebugLog msg showval action = do
+  SAML.logger SAML.Debug $ "entering " ++ msg
+  val <- action
+  let mshowedval = showval val
+  SAML.logger SAML.Debug $ "leaving " ++ msg ++ mconcat [": " ++ fromJust mshowedval | isJust mshowedval]
+  pure val
+
+authorizeIdP :: (HasCallStack, MonadError ServantErr m, Brig.MonadSparToBrig m)
+             => ZUsr -> IdP -> m IdP
+authorizeIdP Nothing _ = throwError err403 { errBody = "Auth token required" }
```

Comment:
This also relates to `type ZUsr = Maybe Brig.UserId`; can we not make this compulsory? It avoids quite a bit of checking in other places

Reply:
It is a run-time error to not have this header set.  A newer version of servant offers to make Headers compulsory on the type level, and then servant would handle the error case for you.

I don't see a significantly simpler way to handle this in the servant version we have, and I don't think it's worth upgrading.

If you disagree: where specifically would you like to handle the `Nothing` case?

Reply:
> If you disagree: where specifically would you like to handle the Nothing case?

Sorry I did not add to the discussion at the time; I find it easier to read and can also allow us to fail earlier. For instance, `idpGet` and `idpDelete` do an unconditional lookup which then fails because the header _may_ not be set. In practice though, this can really only happen in tests since the validation is done by `nginz` which ensures that the access token is valid and the header is set so it's really only theoretical and not in practice but I find it easier to read. Waiting for the newer version of servant is fine of course :)